### PR TITLE
[SYCL] Apply type constraints for accessor::get_multi_ptr() and ++/-- operators of marray

### PR DESCRIPTION
**Pull request by HPS-1**
Addresses two problems:
accessor::get_multi_ptr() should be available only when AccessTarget == target::device.
For marray<...DataT...>, operator++/-- should be available only when DataT != bool.

P.S. There's a typo in Table 57 under section 4.7.6.9 of SYCL 2020 spec (here:https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_buffer_accessor_for_commands):
`global_ptr<access::decorated::legacy> get_pointer() const noexcept` should  actually be
`global_ptr<DataT> get_pointer() const noexcept` as it returns a multi_ptr to the start of this accessor’s underlying buffer. How should I fix this?

Update: Greg has fixed this issue as mentioned in the comment by replacing `global_ptr<access::decorated::legacy> get_pointer()` with `global_ptr<value_type> get_pointer()` in the SYCL spec. I've also updated our code by replacing `global_ptr<DataT> get_pointer()` with `global_ptr<value_type> get_pointer()` accordingly.

### CODE REVIEW

**Comment by bader** [comment]
> There's a typo in Table 57 under section 4.7.6.9 of SYCL 2020 spec (here:https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_buffer_accessor_for_commands):
> `global_ptr<access::decorated::legacy> get_pointer() const noexcept` should actually be
> `global_ptr<DataT> get_pointer() const noexcept` as it returns a multi_ptr to the start of this accessor’s underlying buffer, which is of type DataT. How should I fix this?

@gmlueck, FYI.

**Comment by gmlueck** [comment]
> > There's a typo in Table 57 under section 4.7.6.9 of SYCL 2020 spec (here:https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_buffer_accessor_for_commands):
> > `global_ptr<access::decorated::legacy> get_pointer() const noexcept` should actually be
> > `global_ptr<DataT> get_pointer() const noexcept` as it returns a multi_ptr to the start of this accessor’s underlying buffer, which is of type DataT. How should I fix this?
> 
> @gmlueck, FYI.

Thanks for finding this.  I submitted KhronosGroup/SYCL-Docs#549 to fix the spec.

File: sycl/include/sycl/accessor.hpp [hunk 1]
```diff
@@ -2415,7 +2415,10 @@ class __SYCL_EBO __SYCL_SPECIAL_CLASS __SYCL_TYPE(accessor) accessor :
     return constant_ptr<DataT>(getPointerAdjusted());
   }
 
-  template <access::decorated IsDecorated>
+  template <
+      access::decorated IsDecorated,
+      access::target AccessTarget_ = AccessTarget,
+      typename = std::enable_if_t<(AccessTarget_ == access::target::device)>>
```
[the remark this answers is not in the document]
Comment (steffenlarsen) [reply sycl/include/sycl/accessor.hpp:2421]: I would say we do, yeah. The way we have it defined, `accessor_ptr` was simply a `value_type *`, so the previous call to this was still valid, albeit nonsensical. I think the best option here is to clone the definition here, change the condition to `!= access::target::device` and deprecate that clone.

**Review by bso-intel — APPROVED** [review]

**Comment by AlexeySachkov** [comment]
I think that this PR broke post-commit: https://github.com/intel/llvm/actions/runs/8832557360/job/24250738532 

```
FAIL: SYCL :: ESIMD/lsc/lsc_surf_load_stateless.cpp (654 of 2015)
******************** TEST 'SYCL :: ESIMD/lsc/lsc_surf_load_stateless.cpp' FAILED ********************
Exit Code: 1

Command Output (stdout):
--
# RUN: at line 9
/__w/llvm/llvm/toolchain/bin//clang++   -fsycl -fsycl-targets=spir64 /__w/llvm/llvm/llvm/sycl/test-e2e/ESIMD/lsc/lsc_surf_load_stateless.cpp -fsycl-esimd-force-stateless-mem -o /__w/llvm/llvm/build-e2e/ESIMD/lsc/Output/lsc_surf_load_stateless.cpp.tmp.out
# executed command: /__w/llvm/llvm/toolchain/bin//clang++ -fsycl -fsycl-targets=spir64 /__w/llvm/llvm/llvm/sycl/test-e2e/ESIMD/lsc/lsc_surf_load_stateless.cpp -fsycl-esimd-force-stateless-mem -o /__w/llvm/llvm/build-e2e/ESIMD/lsc/Output/lsc_surf_load_stateless.cpp.tmp.out
# .---command stderr------------
# | In file included from /__w/llvm/llvm/llvm/sycl/test-e2e/ESIMD/lsc/lsc_surf_load_stateless.cpp:17:
# | In file included from /__w/llvm/llvm/llvm/sycl/test-e2e/ESIMD/lsc/Inputs/lsc_surf_load.hpp:9:
# | In file included from /__w/llvm/llvm/toolchain/bin/../include/sycl/ext/intel/esimd.hpp:93:
# | /__w/llvm/llvm/toolchain/bin/../include/sycl/ext/intel/experimental/esimd/memory.hpp:786:7: error: reinterpret_cast from 'pointer_t' (aka 'const __global unsigned int *') to 'unsigned int *' casts away qualifiers
# |   786 |       reinterpret_cast<T *>(acc.get_pointer().get()), offsets, pred);
# |       |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# | /__w/llvm/llvm/toolchain/bin/../include/sycl/ext/intel/experimental/esimd/memory.hpp:807:10: note: in instantiation of function template 
```

`acc.get_pointer()` now accounts for const-ness of the type (depending on access mode) which adds `const` qualifier to source data type in the example above.

Strictly speaking, I think that this is a bug somewhere in ESIMD headers which we exposed. `get_pointer()` is a deprecated API and `get_multi_ptr()` should be used instead, which is also a pointer to `value_type`, i.e. it may have `const` attached to it.

Tagging @v-klochkov for awareness. I will try to take a look at this as well.

A short-term solution would be to do a partial revert, i.e. remove changes we've done around `get_pointer()` return type. The rest of this PR should not be related to this failure.

**Comment by AlexeySachkov** [comment]
> I think that this PR broke post-commit: https://github.com/intel/llvm/actions/runs/8832557360/job/24250738532
> 
> ```
> FAIL: SYCL :: ESIMD/lsc/lsc_surf_load_stateless.cpp (654 of 2015)
> ******************** TEST 'SYCL :: ESIMD/lsc/lsc_surf_load_stateless.cpp' FAILED ********************
> Exit Code: 1
> 
> Command Output (stdout):
> --
> # RUN: at line 9
> /__w/llvm/llvm/toolchain/bin//clang++   -fsycl -fsycl-targets=spir64 /__w/llvm/llvm/llvm/sycl/test-e2e/ESIMD/lsc/lsc_surf_load_stateless.cpp -fsycl-esimd-force-stateless-mem -o /__w/llvm/llvm/build-e2e/ESIMD/lsc/Output/lsc_surf_load_stateless.cpp.tmp.out
> # executed command: /__w/llvm/llvm/toolchain/bin//clang++ -fsycl -fsycl-targets=spir64 /__w/llvm/llvm/llvm/sycl/test-e2e/ESIMD/lsc/lsc_surf_load_stateless.cpp -fsycl-esimd-force-stateless-mem -o /__w/llvm/llvm/build-e2e/ESIMD/lsc/Output/lsc_surf_load_stateless.cpp.tmp.out
> # .---command stderr------------
> # | In file included from /__w/llvm/llvm/llvm/sycl/test-e2e/ESIMD/lsc/lsc_surf_load_stateless.cpp:17:
> # | In file included from /__w/llvm/llvm/llvm/sycl/test-e2e/ESIMD/lsc/Inputs/lsc_surf_load.hpp:9:
> # | In file included from /__w/llvm/llvm/toolchain/bin/../include/sycl/ext/intel/esimd.hpp:93:
> # | /__w/llvm/llvm/toolchain/bin/../include/sycl/ext/intel/experimental/esimd/memory.hpp:786:7: error: reinterpret_cast from 'pointer_t' (aka 'const __global unsigned int *') to 'unsigned int *' casts away qualifiers
> # |   786 |       reinterpret_cast<T *>(acc.get_pointer().get()), offsets, pred);
> # |       |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> # | /__w/llvm/llvm/toolchain/bin/../include/sycl/ext/intel/experimental/esimd/memory.hpp:807:10: note: in instantiation of function template 
> ```
> 
> `acc.get_pointer()` now accounts for const-ness of the type (depending on access mode) which adds `const` qualifier to source data type in the example above.
> 
> Strictly speaking, I think that this is a bug somewhere in ESIMD headers which we exposed. `get_pointer()` is a deprecated API and `get_multi_ptr()` should be used instead, which is also a pointer to `value_type`, i.e. it may have `const` attached to it.
> 
> Tagging @v-klochkov for awareness. I will try to take a look at this as well.
> 
> A short-term solution would be to do a partial revert, i.e. remove changes we've done around `get_pointer()` return type. The rest of this PR should not be related to this failure.

I'm working on a fix in #13567

**Pull request merged**